File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <iostream>

using namespace std;
vector<vector<int>> candidate;

unordered_set<int> unit;
void makeUnit(int num, int id, int size){
    if(id == size){
        unit.insert(num); 
        return;
    }
    vector<int> candi = candidate[id];
    for(int i = 0; i < candi.size(); i++){ // candi 중 하나 고르면 됨. 이전에 안나왔던 것중에
        if((num & (1 << candi[i])) != (1 << candi[i])){
            makeUnit(num|(1 << candi[i]), id+1, size);
        }
    }
}


int solution(vector<string> user_id, vector<string> banned_id) {
    // 각 user에 번호 매기기. 0번~
    unordered_map<string, int> user; 
    for(int i = 0; i < user_id.size(); i++){
        user[user_id[i]] = i;
    }
    
    // 각 banned_id가 가능한 id 조합 구하기 -> vector에 넣기
    candidate.assign(banned_id.size(), vector<int>());
    for(int j = 0; j < banned_id.size(); j++){
        string b = banned_id[j];
        // 유저 돌며 맞는 id 찾아서 번호 넣기
        for(auto& u: user_id){
            bool avail = true;
            if(u.size() == b.size()){
                for(int i = 0; i < u.size(); i++){
                    if(b[i] != '*' && b[i] != u[i]){
                        avail = false; break;
                    }
                }
                if(avail) {
                    candidate[j].push_back(user[u]);
                }
            }
        }
    }
    
    // 하나씩 돌면서 기존의 id와 겹치지 않는 id 선택 -> set에 넣기
    makeUnit(0, 0, banned_id.size());
    
    // 당첨에서 제외되어야 할 제재 아이디 목록은 몇가지 경우의 수가 가능한 지
    return unit.size();
}

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.java
import java.util.*; 

class Solution {
    
    HashMap<Integer, List<Integer>> available = new HashMap<>(); 
    HashSet<Integer> resultMasks = new HashSet<>(); 
    
    // 당첨에서 제외되어야 할 제재 아이디 목록은 몇가지 경우의 수가 가능한 지 return
    public int solution(String[] user_id, String[] banned_id) {

        
        // 불량 사용자 순회하면서 어떤게 가능한지 확인
        for(int b = 0; b < banned_id.length; b++){ // 8
            String banned = banned_id[b]; 
            for(int u = 0; u < user_id.length; u++){ // 8
                String user = user_id[u]; boolean avail = true;
                // 적합한지
                if(user.length() != banned.length()) continue; 
                for(int i = 0; i < user.length(); i++){ // 8
                    if(banned.charAt(i) != '*' && user.charAt(i) != banned.charAt(i)) {
                        avail = false; break; 
                    }
                }
                if(avail) {
                    available.computeIfAbsent(b, k -> new ArrayList<>()).add(u);
                    // System.out.println(banned + " " + user);
                }
            }
        }
        
        // 하나씩 고르는데 중복 안되도록 (비트마스킹 사용)
        dfs(0, 0, banned_id.length);
       
        
        return resultMasks.size();
    }
    
     public void dfs(int bannedIndex, int mask, int bannedCount) {
        // 모든 불량 사용자 패턴에 사용자 배정 완료
        if (bannedIndex == bannedCount) {
            resultMasks.add(mask);
            return;
        }

        for (int userIndex : available.get(bannedIndex)) {
            int bit = 1 << userIndex;

            // 이미 선택한 사용자
            if ((mask & bit) != 0) continue;

            dfs(bannedIndex + 1, mask | bit, bannedCount);
        }
    }
}